demos/minimal/actors/sample_processor.py

```diff
@@ -1,63 +1,67 @@
 from improv.actor import Actor
-import numpy as np
+from queue import Empty
 import logging
+import zmq
+import numpy as np
 
 logger = logging.getLogger(__name__)
 logger.setLevel(logging.INFO)
 
 
 class Processor(Actor):
-    """Sample processor used to calculate the average of an array of integers.
-
-    Intended for use with sample_generator.py.
+    """
+    Process data and send it through zmq to be visualized.
     """
 
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
 
     def setup(self):
-        """Initializes all class variables.
-
-        self.name (string): name of the actor.
-        self.frame (ObjectID): StoreInterface object id referencing data from the store.
-        self.avg_list (list): list that contains averages of individual vectors.
-        self.frame_num (int): index of current frame.
         """
-
+        Creates and binds the socket for zmq.
+        """
         self.name = "Processor"
-        self.frame = None
-        self.avg_list = []
-        self.frame_num = 1
+
+        context = zmq.Context()
+        self.socket = context.socket(zmq.PUB)
+        self.socket.bind("tcp://127.0.0.1:5555")
+
         logger.info("Completed setup for Processor")
 
     def stop(self):
-        """Trivial stop function for testing purposes."""
-
         logger.info("Processor stopping")
+        self.socket.close()
+        return 0
 
     def runStep(self):
-        """Gets from the input queue and calculates the average.
-
-        Receives an ObjectID, references data in the store using that
-        ObjectID, calculates the average of that data, and finally prints
-        to stdout.
         """
-
-        frame = None
+        Receives data from the queue, prepares 2D data with x and y coordinates,
+        and sends it through the socket along with the frame number.
+        """
         try:
-            frame = self.q_in.get(timeout=0.001)
-
-        except:
-            logger.error("Could not get frame!")
-            pass
-
-        if frame is not None and self.frame_num is not None:
-            self.done = False
-            self.frame = self.client.getID(frame[0][0])
-            avg = np.mean(self.frame[0])
-
-            # print(f"Average: {avg}")
-            self.avg_list.append(avg)
-            # print(f"Overall Average: {np.mean(self.avg_list)}")
-            # print(f"Frame number: {self.frame_num}")
-            self.frame_num += 1
+            # Retrieve data from the queue
+            frame = self.q_in.get(timeout=0.05)
+        except Empty:
+            return  # No data received, skip this step
+        except Exception as e:
+            logger.error(f"Error retrieving frame: {e}")
+            return
+
+        if frame is not None:
+            try:
+                # Fetch the dictionary from the data store
+                data_dict = self.client.getID(frame[0][0])
+
```

Comment:
See comment in generator, but once you fix things there you should be able to fix the code below to unpack from a numpy array instead of a dictionary.

Reply:
You should also check out the minimal demo sample processor in the dev branch and mimic the code there.